DOC: Clarify sorting and order of categoricals and fix typo

StefRe opened the pull request:
Without the addition in the text it‘s unclear, what `s.sort_values(inplace=True)`
should demonstrate. Show some meaningful output now. Also, when using `astype` to
contrast ordered/unordered behavior, I think it‘s clearer to use the existing,
unordered Series than to create a new one.

Fix minor typo in categorical.py.


- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

StefRe committed:
DOC: Clarify sorting and order of categoricals and fix typo

Without the addition in the text it‘s unclear, what s.sort_values(inplace=True)
should demonstrate. Show some meaningful output now. Also, when using astype to
contrast ordered/unordered behavior, I think it‘s clearer to use the existing,
unordered Series than to create a new one.

mroeschke commented on doc/source/user_guide/categorical.rst:
```diff
@@ -439,15 +439,19 @@ Sorting and order
+    s.sort_values(ascending=False)
+    try:
+        s.min()
+    except TypeError as e:
```
It would better to demonstrate this with the `:okexcept:` directive

StefRe replied:
Hm, if I look at examples where the `:okexcept:` directive is used, for instance [here](https://pandas.pydata.org/docs/dev/user_guide/gotchas.html#using-if-truth-statements-with-pandas) or [here](https://pandas.pydata.org/docs/dev/user_guide/advanced.html#renaming-names-of-an-index-or-multiindex), I'm not sure if this is so good an idea in the given case, as it inserts quite a lengthy (~15 lines) traceback which, in my opinion, distracts from reading and makes the example less clear than with `try` / `except`.

mroeschke replied:
Could you show an image of how the docs would render if `:okexcept:` was used?

StefRe replied:
see the links I provided in the reply:
```
.. ipython:: python
    :okexcept:

    if pd.Series([False, True, False]):
        print("I was true")
```
renders as 
![grafik](https://user-images.githubusercontent.com/19879328/180955288-b60e6b5c-9687-4197-a6e5-94f5486710bc.png)

and
```
.. ipython:: python
   :okexcept:

   mi.levels[0].name = "name via level"
```
renders as
![grafik](https://user-images.githubusercontent.com/19879328/180955412-2c29d20f-94f6-4407-acbb-59256563f7dc.png)

mroeschke replied:
Sorry to clarify, how would the docs you modified in this PR render with `:okexcept:`?

StefRe replied:
```
.. ipython:: python
    :okexcept:

    s = pd.Series(pd.Categorical(["a", "b", "c", "a"], ordered=False))
    s.sort_values(ascending=False)
    s.min()
```
results in
![out](https://user-images.githubusercontent.com/19879328/183150140-73b440d6-f7b5-4b7b-9787-e3341a652c5f.png)

mroeschke replied:
I would still be partial to adding `:okexcept:` here. Though verbose, it shows the exact behavior that the user should see.

StefRe replied:
I'm afraid I can't agree here: IMO the purpose here is just to demonstrate that an exception will be raised when trying to do `min` or `max` on a not ordered categorical series. If we then catch it or let it be thrown doesn't play a role for this purpose, so I'd prefer the visually less distracting option of catching it. This is the way it's done all the way in [the given section of the user guide](https://pandas.pydata.org/docs/dev/user_guide/categorical.html) (9 times), why should we treat this 10th time differently? Maybe we can leave the `try/except` here just for uniformity's sake.

mroeschke replied:
Personally, I think those other sections should also use `:okexcept:` as well.

mroeschke commented on doc/source/user_guide/categorical.rst:
```diff
@@ -439,15 +439,19 @@ Sorting and order
+    except TypeError as e:
+        print(f"TypeError: {e}")
+    s = s.astype(CategoricalDtype(ordered=True))
+    s.sort_values()
```
I think this will need to be `s = s.sort_values()`

StefRe replied:
Of course you could write `s = s.sort_values()` followed by `s` to show the result but as we don't need the sorted series further down (I mean for the following code it doesn't play a role if `s` has been sorted or not) I think this could be shortened to just `s.sort_values()` which will output the returned sorted Series.  

In principle, one could remove this line completely, because if we can sort an unordered categories series, we can sort an ordered categories series all the more (it's a leftover from the beginning, when sorting of an unordered categories series was not allowed, this was changed in 87fec5b527dcbb).

mroeschke replied:
Sure removing this line would be better then

StefRe commented:
Oh. sorry - I misread your question. I don't have access to my computer
now, will reply at end of next week.

Matthew Roeschke ***@***.***> schrieb am Di., 26. Juli 2022,
18:44:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In doc/source/user_guide/categorical.rst
> <https://github.com/pandas-dev/pandas/pull/47846#discussion_r930197755>:
>
> >
>  .. ipython:: python
>
>      s = pd.Series(pd.Categorical(["a", "b", "c", "a"], ordered=False))
> -    s.sort_values(inplace=True)
> -    s = pd.Series(["a", "b", "c", "a"]).astype(CategoricalDtype(ordered=True))
> -    s.sort_values(inplace=True)
> -    s
> +    s.sort_values(ascending=False)
> +    try:
> +        s.min()
> +    except TypeError as e:
>
> Sorry to clarify, how would the docs you modified in this PR render with
> :okexcept:?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/47846#discussion_r930197755>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEXVLIA53J6MBCNVV725PVTVWAIVRANCNFSM54RYFBTQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

datapythonista reviewed (commented):
@StefRe what do you think if you implement the suggested changes here, and then you open an issue to discuss the problem of too long tracebacks. I totally see your point on that, and would be great to not have these huge tracebacks in the middle of the documentation. But let's better discuss that separately, maybe there are better solutions than using try/except blocks, which aren't great either. Thanks!

mroeschke commented:
xref https://github.com/pandas-dev/pandas/issues/10715 for a discussion on limiting tracebacks.

I do sympathize and agree that having very long tracebacks in the documentation can be distract, but my primary objection with `try/except` in an ad-hoc fashion is that it can over-obscure the expected exception behavior.

StefRe commented:
@datapythonista Sounds good in principle but I just don't want to be the guy who insanely blew up the documentation by adding estimated 400 to 500 lines of traceback, making it practically unreadable.   

So maybe we can do it just the other way round: introduce this one additional try/execpt to the 9 already existing in this tutorial and open another issue about replacing try/except constructs by (preferably shortened) tracebacks (there is a small number of try/expcepts in other pandas tutorials too). What do you think about this, @mroeschke ?

mroeschke commented:
I would slightly prefer https://github.com/pandas-dev/pandas/issues/10715 addressed before adding more `try/except` patterns such that the other `try/except` patterns could be eliminated too.

StefRe commented:
just for myself a short summary of previous activity on this:

- First try: 
#10727
worked locally but broke CI
- Reversal:
#10916
reverts the above
- Second try:
#10925
didn't work either
- Upstream:
ipython/ipython#8686
more or less the same as #10727 but upstream

It's not mentioned what exactly was the error on CI (https://github.com/pandas-dev/pandas/pull/10727#issuecomment-133594091 just states "doc builds are stopping since this commit").

Today, pandas uses the upstream IPython extension instead of its local copy (#19657), so we'd
1. make a PR upstream
2. verify that pandas docs build locally with this PR
3. somehow use this updated upstream version for building the docs in CI and see if it still works. 

This last part (# 3) is not clear to me how to accomplish - any thoughts about it?

datapythonista commented:
Limiting the tracebacks doesn't necessarily need to be addressed via ipython. We can also use other code directives in sphinx where we specify the output instead of being autogenerated, or maybe the python interpreter itself can be configured or monkeypatched to change the traceback behaviour.

Maybe for now you can explain in the docs about the exception instead of showing it, so we don't have the try/except or the long traceback.

StefRe commented:
> We can also use other code directives in sphinx where we specify the output instead of being autogenerated

certainly one could use just code blocks (`.. code-block:: ipython`) as done [here](https://pandas.pydata.org/docs/dev/user_guide/basics.html#comparing-array-like-objects) but in my opinion this is worse than try/except (in any case you'll need to manually adjust the command numbers to avoid such glitches where after # 68 come # 55 and 56 and then # 69:
![image](https://user-images.githubusercontent.com/19879328/188283161-5ccf09ac-4267-453b-a7c3-6f9b35fff44d.png)

<br>

My idea was more to change IPython's exception handling by [setting](https://ipython.readthedocs.io/en/stable/api/generated/IPython.core.interactiveshell.html#IPython.core.interactiveshell.InteractiveShell.set_custom_exc) a custom handler. This is rather straightforward, the only thing I'm still struggling with is how to completely undo this and revert it to the default. This would also cover cases where the error message stretches over more than one line (as in the concrete case here).

StefRe commented:
Converted to DRAFT until #10715 is resolved.  

(I added an optional `no_traceback` argument to the `:okexcept:` option, works fine locally, will test it on CI next week:
![image](https://user-images.githubusercontent.com/19879328/188332749-e94c6a8b-a045-453e-9151-4facfc7b208c.png))

mroeschke commented:
Thanks for looking into also limiting the tracebacks @StefRe!

StefRe committed:
Merge branch 'main' into DOC/categorical

StefRe commented:
> respond to this comment if you're still interested in working on this.

still interested but IPython seem to ignore PR at the moment (no reaction whatsoever for a month already)

simonjayhawkins commented:
> > respond to this comment if you're still interested in working on this.
> 
> still interested but IPython seem to ignore PR at the moment (no reaction whatsoever for a month already)

ok let's close for now to clear the queue.
